### ISSUE TITLE
Error with StrEnum instance as dict key

### ISSUE THREAD

**Issue by Kxnr** [ijl/orjson#414]
## Problem Description

When using a `StrEnum` or an enum with `str` as a base class as a key in a dictionary. we receive a `TypeError`.

## Observed Behavior

Minimal example for python 3.11+. The same behavior can be observed in earlier python versions using an enum defined as `class A(str, Enum)`.

```python
from enum import StrEnum
import orjson

class A(StrEnum):
    a = "abc"

test = {A.a: 1234}

orjson.dumps(test)
```

results in:
```
TypeError: Dict key must be str
```

## Expected Behavior

Members of a `StrEnum` are string instances--`isinstance(A.a, str) == True`--and thus seem like they should be valid dictionary keys. This issue can be worked around using the `OPT_NON_STR_KEYS` flag, but that allows a much more permissive set of keys which is undesirable.

**Comment by Kxnr** [comment]
I think your example may demonstrate the opposite of what you expect it to:

```python
In [1]: from enum import StrEnum
   ...: import orjson
   ...:
   ...: class A(StrEnum):
   ...:     a = "abc"
   ...:
   ...: test = {A.a: 1234}

In [2]: test[A.a] == test["abc"]
Out[2]: True

In [3]: hash(A.a) == hash("abc")
Out[3]: True

In [4]: A.a == "abc"
Out[4]: True
```

Yes, `StrEnum` members are a subclass of `str`. Expecting them to act differently than a `str` would be the surprising behavior, imo, 
They are, in every meaningful sense I'm aware of, equivalent to the string they wrap.

**Comment by ddorian** [comment]
@Kxnr did you find a workaround?

**Comment by Kxnr** [comment]
@ddorian no, not aside from the `orjson.OPT_NON_STR_KEYS` mentioned in the original report. We've dropped this library for serialization as a result and are using `pydantic` models.

**Comment by camillol** [comment]
@ijl 
> I agree you could potentially check for a StrEnum before falling back, but then because they are different types and will always have different hash values you can do this:
> 
> In  [8]: orjson.dumps({A.a: 1, "abc": 2}, option=orjson.OPT_NON_STR_KEYS)
> Out[8]: b'{"abc":2}'
> In  [9]: orjson.dumps({"abc": 2, A.a: 1}, option=orjson.OPT_NON_STR_KEYS)
> Out[9]: b'{"abc":1}'
> And that is some surprising behavior deserving the warning and opt-in of `OPT_NON_STR_KEYS`.

`StrEnum` is specifically designed so that its members are strings and can be used as such. The surprising behavior in your example is actually the expected behavior of `StrEnum`, and occurs regardless of whether `orjson` is involved.

> StrEnum is the same as [Enum](https://docs.python.org/3/library/enum.html#enum.Enum), but its members are also strings and can be used in most of the same places that a string can be used.
[0]: https://docs.python.org/3/library/enum.html#enum.StrEnum

A StrEnum member is equal to the corresponding string, and its hash is also the same:

```python
In [1]: from enum import StrEnum
   ...: import orjson
   ...: class Color(StrEnum):
   ...:     RED = "red"
   ...:     GREEN = "green"
   ...: 

In [2]: Color.RED == "red"
Out[2]: True

In [3]: hash(Color.RED) == hash("red")
Out[3]: True
```

Therefore, in a dictionary they are considered the same key. The behavior you're seeing with a dictionary using both keys actually exists independently of orjson, and it's the same behavior as specifying the same key more than once in a dictionary. The dictionary keeps the first version of the key it sees, and the last value:

```python
In [4]: {Color.RED:1, "red":2}
Out[4]: {<Color.RED: 'red'>: 2}

In [5]: {"red":1, Color.RED:2}
Out[5]: {'red': 2}

In [6]: {Color.RED:1, "red":2} == {"red":1, Color.RED:2}
Out[6]: True
```

So, in your example, orjson was receiving a dictionary with a single entry.
If you dump a dictionary with a StrEnum key using `json`, it gets serialized as a string:

```python
In [7]: import json

In [8]: json.dumps({Color.RED:1})
Out[8]: '{"red": 1}'

In [9]: json.loads(json.dumps({Color.RED:1}))
Out[9]: {'red': 1}
```

Of course, it will get deserialized as a string too. This is the expected behavior with StrEnum instances, not just because of `json`'s precedent, but because it's what the `StrEnum` documentation says:

> The result of any string operation performed on or with a StrEnum member is not part of the enumeration.

Likewise:
```python
In [10]: Color.RED + ""
Out[10]: 'red'
```

So it's not surprising to StrEnum users that converting to json and back would yield a simple string.

I understand the rationale for orjson to be cautious and not silently convert arbitrary `str` subclasses to plain strings. But given `StrEnum`'s design intention and documented behavior, I think orjson should conform to json's behavior and accept it as a string key without requiring `orjson.OPT_NON_STR_KEYS`.
